zeus/optimizer/batch_size/server/batch_size_state/repository.py

```diff
@@ -1,126 +1,116 @@
 from zeus.optimizer.batch_size.server.batch_size_state.commands import (
-    CreateExploration,
-    UpdateExploration,
+    CreateTrial,
+    ReadTrial,
+    UpdateTrial,
 )
 from zeus.optimizer.batch_size.server.batch_size_state.models import (
     BatchSizeBase,
-    ExplorationsPerBs,
     ExplorationsPerJob,
-    ExplorationStateModel,
-    GaussianTsArmStateModel,
-    MeasurementOfBs,
-    MeasurementsPerBs,
+    GaussianTsArmState,
+    Trial,
+    TrialResult,
+    TrialResultsPerBs,
 )
 from zeus.optimizer.batch_size.server.database.repository import DatabaseRepository
 from zeus.optimizer.batch_size.server.database.schema import (
-    ExplorationState,
-    GaussianTsArmState,
-    Measurement,
+    GaussianTsArmStateTable,
+    TrialStatus,
+    TrialTable,
+    TrialType,
 )
+from zeus.optimizer.batch_size.server.exceptions import ZeusBSOValueError
+from zeus.util.logging import get_logger
+
+logger = get_logger(__name__)
 
 
 class BatchSizeStateRepository(DatabaseRepository):
     """Repository for handling batch size related operations."""
 
-    async def get_measurements_of_bs(
-        self, batch_size: BatchSizeBase, window_size: int
-    ) -> MeasurementsPerBs:
-        """Load window size amount of measurement for a given batch size. If window size <= 0, load all measurements.
-
-        Args:
-            batch_size (BatchSizeBase): The batch size object.
-            window_size (int): The size of the measurement window.
+    def __init__(self, session: AsyncSession):
+        """Set db session and intialize fetched trial. We are only updating one trial per session."""
+        super().__init__(session)
+        self.fetched_trial: Trial | None = None
+        self.fetched_arm: GaussianTsArmState | None = None
 
-        Returns:
-            MeasurementsPerBs: Measurements for the given batch size.
-        """
+    async def get_next_trial_number(self, job_id: str) -> int:
+        """Get next trial number of a given job. Trial number starts from 1 and increase by 1 at a time."""
         stmt = (
-            select(Measurement)
+            select(TrialTable)
             .where(
                 and_(
-                    Measurement.job_id == batch_size.job_id,
-                    Measurement.batch_size == batch_size.batch_size,
+                    TrialTable.job_id == job_id,
                 )
             )
-            .order_by(Measurement.timestamp.desc())
+            .order_by(TrialTable.trial_number.desc())
+            .limit(1)
         )
-        if window_size > 0:
-            stmt = stmt.limit(window_size)
-
         res = (await self.session.scalars(stmt)).all()
-        return MeasurementsPerBs(
-            job_id=batch_size.job_id,
-            batch_size=batch_size.batch_size,
-            measurements=[MeasurementOfBs.from_orm(m) for m in res],
-        )
+        if len(res) == 0:
+            return 1
+        else:
+            return res[0].trial_number + 1
```

Comment:
I thought `COUNT` would iterate all results and count a number of pks, but `order_by` + `limit` will only fetch one-row using indexing. Maybe I'm wrong.  Actually I gotta add an index to `trial_number`

Reply:
Ah missed the limit(1). But still, I think sort then limit will make the DB (internally) find all the rows, sort, and return the first one. In contrast, with count, it will find all the nodes and count then, saving us a nlogn sort operation.

Reply:
Ah I missed that `where` is getting evaluated first. I did some digging and ppl said `MAX` should be faster than `COUNT`. Should be a minor advantage though.

Reply:
Yeah I'm fine with either as long as we're not making the DB sort.